packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -147,19 +147,21 @@ class ConnectionStateCatchup extends ConnectionStateHandlerPassThrough {
         switch (value) {
             case ConnectionState.Connected:
                 assert(this._connectionState === ConnectionState.CatchingUp, 0x3e1 /* connectivity transitions */);
-                assert(this.catchUpMonitor !== undefined,
-                    0x3e2 /* catchUpMonitor should always be set if pendingClientId is set */);
-                this.catchUpMonitor.on("caughtUp", this.transitionToConnectedState);
+                // Create catch-up monitor here (not earlier), as we might get more exact info by now about how far
+                // client is behind through join signal. This is only true if base layer uses signals (i.e. audience,
+                // not quorum, including for "rea" connections) to make decisions about moving to "connected" state.
+                // In addition to that, in its current form, doing this in ConnectionState.CatchingUp is dangerous as
+                // we might get callback right away, and it will screw up state transition (as code outside of switch
+                // statement will overwrite current state).
+                assert(this.catchUpMonitor === undefined, "catchUpMonitor should be gone");
+                this.catchUpMonitor = new CatchUpMonitor(this.deltaManager, this.transitionToConnectedState);
                 return;
             case ConnectionState.Disconnected:
                 this.catchUpMonitor?.dispose();
                 this.catchUpMonitor = undefined;
                 break;
             case ConnectionState.CatchingUp:
                 assert(this._connectionState === ConnectionState.Disconnected, 0x3e3 /* connectivity transitions */);
```

Comment:
This doesn't feel relevant to this layer

Reply:
Ohh nevermind, I keep think in terms of inheritance not pipeline.  This is good.  If we get more layers, maybe there's a shared "validateTransition" function that everyone calls at the top of their `connectionStateChanged` or something.